## <center>Carregue dados
### Os datasets, geralmente, contêm diferentes tipos de dados e podem ter nomes associados às linhas ou às colunas
### Por isso, eles, normalmente, são melhor acomodados usando um DataFrame
### A biblioteca **pandas** possui o tipo DataFrame

In [4]:
import pandas as pd

auto = pd.read_csv('Auto.csv') 

## <center>Escreva o conteúdo da variável Auto</center>

In [5]:
auto

mpg  cylinders  displacement horsepower  weight  acceleration  year  \
0    18.0          8         307.0        130    3504          12.0    70   
1    15.0          8         350.0        165    3693          11.5    70   
2    18.0          8         318.0        150    3436          11.0    70   
3    16.0          8         304.0        150    3433          12.0    70   
4    17.0          8         302.0        140    3449          10.5    70   
..    ...        ...           ...        ...     ...           ...   ...   
392  27.0          4         140.0         86    2790          15.6    82   
393  44.0          4          97.0         52    2130          24.6    82   
394  32.0          4         135.0         84    2295          11.6    82   
395  28.0          4         120.0         79    2625          18.6    82   
396  31.0          4         119.0         82    2720          19.4    82   

     origin                       name  
0         1  chevrolet chevelle malibu  
1         1          buick skylark 320  
2         1         plymouth satellite  
3         1              amc rebel sst  
4         1                ford torino  
..      ...                        ...  
392       1            ford mustang gl  
393       2                  vw pickup  
394       1              dodge rampage  
395       1                ford ranger  
396       1                 chevy s-10  

[397 rows x 9 columns]

## <center>O tipo da variável Auto é DataFrame - confira usando a função type()</center>

In [6]:
type(auto)

pandas.core.frame.DataFrame

## <center>Exiba coluna horsepower</center>

In [7]:
auto['horsepower']

0      130
1      165
2      150
3      150
4      140
      ... 
392     86
393     52
394     84
395     79
396     82
Name: horsepower, Length: 397, dtype: object

### Na linha final, consta o nome da coluna - horsepower, a quantidade de dados - 397 e o tipo (dtype) destes dados - object. O tipo <u>object</u> indica que os valores foram interpretados como strings. Para descobrir o motivo, vamos ver os valores únicos com o método unique() do pacote numpy (Numerical Python)
## <center>Importe o numpy</center>

In [8]:
import numpy as np

## <center>Exiba os valores únicos da coluna horsepower com o método np.unique()</center>

In [9]:
np.unique(auto['horsepower'])

array(['100', '102', '103', '105', '107', '108', '110', '112', '113',
       '115', '116', '120', '122', '125', '129', '130', '132', '133',
       '135', '137', '138', '139', '140', '142', '145', '148', '149',
       '150', '152', '153', '155', '158', '160', '165', '167', '170',
       '175', '180', '190', '193', '198', '200', '208', '210', '215',
       '220', '225', '230', '46', '48', '49', '52', '53', '54', '58',
       '60', '61', '62', '63', '64', '65', '66', '67', '68', '69', '70',
       '71', '72', '74', '75', '76', '77', '78', '79', '80', '81', '82',
       '83', '84', '85', '86', '87', '88', '89', '90', '91', '92', '93',
       '94', '95', '96', '97', '98', '?'], dtype=object)

* Vemos que o tipo é object por causa do **?**, que está sendo usado para codificar valores ausentes.
* Para resolver o problema, devemos fornecer o argumento na_values ao pd.read_csv().
* Com isso, cada ocorrência de **?** no arquivo é substituída pelo valor np.nan, o que significa que não é um número

## <center> Antes de substituir o ?, some os valores da coluna horsepower <br> utilizando o método sum() da Series<br><br>Cada coluna de um DataFrame no pandas é <br>uma estrutura de dados do tipo Series</center>

In [12]:
auto['horsepower'].sum()

'13016515015014019822021522519017016015022595959785884687909511390215200210193889095?1001051008810016517515315018017017511072100888690707665696070958054908616517515015315020815516019097150130140150112768769869297808817515014513715019815015815021522517510510010088954615016717018010088729490851079014523049759111215011012218095?100100678065751001101051401501501401508367785261757575979367951057272170145150148110105110951101101297583100789671979770909588981155386819279831401501201521001058190526070531007811095717075721021508810812018014513015068805896701451101451301101051009818017019014978887589638367789711011048665270601101401391059585881009010585110120145165139140689597759510585971031251151337168115858890110130129138135155142125150716580807712571907070656990115115907660706590889090789075927565105654848676767?676213210088?72848492110845864606765626863656574?7575100748076116120110105888588888885849092746868637088757067676711085921129684908652847982'

**Resultado:** sum() concatenou strings 

In [11]:
auto['horsepower'].value_counts()['?']

5

## <center> Substitua o valor de '?' utilizando o keyword argumento 'na_values' <br> do método pd.read_csv</center>

In [14]:
auto = pd.read_csv('Auto.csv',na_values=['?'])

## <center> Some novamente</center>

In [15]:
auto['horsepower'].sum()

40952.0

**<center>Parou aqui**